docstring_parser/parser/google.py

```diff
@@ -2,27 +2,87 @@
 
 import inspect
 import re
+import typing as T
+from collections import namedtuple
+from enum import IntEnum
 
 from .common import Docstring, DocstringMeta, ParseError
 
-_sections = {
-    "Arguments": "param",
-    "Args": "param",
-    "Parameters": "param",
-    "Params": "param",
-    "Raises": "raises",
-    "Exceptions": "raises",
-    "Except": "raises",
-    "Attributes": None,
-    "Example": None,
-    "Examples": None,
-    "Returns": "returns",
-    "Yields": "yields",
-}
-_titles_re = re.compile(
-    "^(" + "|".join("(%s)" % t for t in _sections) + "):", flags=re.M
-)
-_valid = {t for t, a in _sections.items() if a}
+
+class SectionType(IntEnum):
+    """Types of sections."""
+
+    SINGULAR = 0
+    """For sections like examples."""
+
+    MULTIPLE = 1
+    """For sections like params."""
+
+    SINGULAR_OR_MULTIPLE = 2
+    """For sections like returns or yields."""
+
+
+Section = namedtuple("Section", "title key type")
+
+
+class _globals:
+    sections: T.Dict[str, Section] = None
+    title_colon: bool = None
+    titles_re: re.Pattern = None
+    valid: T.Set[str] = None
```

Comment:
You initialize these to `None` so the type hints should really be `T.Optional`.

Reply:
Let's refactor it a bit:

```python3
class Config:
    def __init__(self, sections: T.Dict[str, Section], title_colon: bool) -> None:
        self.sections = sections
        self.title_colon = title_colon
        self.titles_re = …
        self.valid = …


config: T.Optional[Config] = None


def setup(sections: T.Optional[T.List[Section]] = None, title_colon: bool = True) -> None:
    global config

    blabla

    config = Config(
        sections=…,
        title_colon=…,
    )


setup()
```

and further later than using `_globals`, use `config`

Reply:
Refactored with classes because it was becoming weird.